src/GraphQL/Execution/SubscriptionExecutionStrategy.cs

```diff
@@ -0,0 +1,296 @@
+        /// </summary>
+        public SubscriptionExecutionStrategy()
+            : this(new ParallelExecutionStrategy())
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance with the specified execution strategy for child nodes.
+        /// </summary>
+        public SubscriptionExecutionStrategy(IExecutionStrategy baseExecutionStrategy)
+        {
+            _baseExecutionStrategy = baseExecutionStrategy ?? throw new ArgumentNullException(nameof(baseExecutionStrategy));
+        }
+
+        /// <summary>
+        /// Gets a static instance of <see cref="SubscriptionExecutionStrategy"/>.
+        /// </summary>
+        public static SubscriptionExecutionStrategy Instance { get; } = new();
+
+        /// <summary>
+        /// Executes a GraphQL subscription request and returns the result. The result consists
+        /// of one or more streams of GraphQL responses, returned within <see cref="SubscriptionExecutionResult.Streams"/>.
+        /// No serializable <see cref="ExecutionResult"/> is directly returned.
+        /// </summary>
+        public override async Task<ExecutionResult> ExecuteAsync(ExecutionContext context)
+        {
+            var rootType = GetOperationRootType(context);
+            var rootNode = BuildExecutionRootNode(context, rootType);
+
+            var streams = await ExecuteSubscriptionNodesAsync(context, rootNode.SubFields!).ConfigureAwait(false);
+
+            // if execution is successful, errors and extensions are not returned per the graphql-ws protocol
+            // if execution is unsuccessful, the DocumentExecuter will add context errors to the result
+
+            return new SubscriptionExecutionResult(context)
+            {
+                Executed = true,
+                Streams = streams,
+            };
+        }
+
+        private async Task<IDictionary<string, IObservable<ExecutionResult>>?> ExecuteSubscriptionNodesAsync(ExecutionContext context, ExecutionNode[] nodes)
+        {
+            var streams = new Dictionary<string, IObservable<ExecutionResult>>();
+
+            foreach (var node in nodes)
+            {
+                var stream = await ResolveEventStreamAsync(context, node).ConfigureAwait(false);
+                if (stream == null)
+                    return null;
+                streams[node.Name!] = stream;
+            }
+
+            return streams;
+        }
+
+        /// <summary>
+        /// Asynchronously returns a stream of <see cref="ExecutionResult"/> responses for the
+        /// specified <see cref="ExecutionNode"/>.
+        /// </summary>
+        protected virtual async Task<IObservable<ExecutionResult>?> ResolveEventStreamAsync(ExecutionContext context, ExecutionNode node)
+        {
+            context.CancellationToken.ThrowIfCancellationRequested();
+
+            var resolveContext = new ReadonlyResolveFieldContext(node, context);
+
+            IObservable<object?> subscription;
```

Comment:
Hmm. Now, when I read the documentation about Rx and some docs about subscriptions it became clear to me what confused me for a long time. I would like to propose some renamings here to align terms with ones in GraphQL spec and reactive stuff.
1. IObservable<object?> subscription -> IObservable<object?> sourceStream (because _subscription_ is usually a result of subscribing **to** observable but not an observable **itself**)
2. https://spec.graphql.org/October2021/#sec-Subscription . I would like to make clear separation for _source stream_ as observable of source values from resolver and _response stream_ as observable of execution results. This also implies renaming `IEventStreamResolver`. From reading the spec `EventStream` means sutable for both streams - source or response so it confuses. Conseptually `IEventStreamResolver` may be somehow (`ISourceStreamResolver` ?) named to emphasize _source_ nature of its observable. In contrast `ResolveEventStreamAsync(ExecutionContext context, ExecutionNode node)` method here may be named as `ResolveResponseStreamAsync` to emphasize _response_ nature of its observable after transforming/executing source value into `ExecutionResult`.

Reply:
Sounds good, although if it's just renaming, perhaps we should merge this first (after finishing review) before another PR to do renaming.  But in general, your suggestions sound good.

Reply:
For me it would be better to do all this stuff here - personally I understand/feel the code better if it has good naming. One more suggestion after renaming `IEventStreamResolver` to `ISourceStreamResolver` : rename `SubscribeAsync` to `CreateSourceStream`. I think this is obvious change now and better reflects the nature of method. Method does nothing with _subscription_ itself but creates a object (stream of events) that one may use to _subscribe to_.

Reply:
I fear this will be a very large quantity of changes if we try to properly change all of the names that relate to subscriptions within the codebase.  Not only are we talking about changing some interface names, but also parameter names for field builders and similar also needs to be changed.  It just seems like this PR has enough changes in it.

![image](https://user-images.githubusercontent.com/6377684/159187276-7c8c65ff-22f6-4393-b62a-460020c5f894.png)

![image](https://user-images.githubusercontent.com/6377684/159187283-afe70d74-4f89-4ef6-872e-914091ed3be1.png)

Reply:
Well, OK. Let's change it after. I believe it is worth to do. I myself was difficult to figure out what was written and what was meant in subscriptions internals.